File: randology-dump-master/thesis/src/uk/ac/stir/randology/generator/MWC.java
package uk.ac.stir.randology.generator;

/**
 * An implementation of the Multiply-With-Carry generator from Numerical Recipes.
 * @author zak
 */
public class MWC implements RNG {
	
	private long multiplier = 0xffffda61L;
	private long seed;
	
	public MWC(long seed) {
		this.seed = seed;
	}
	
	public String getName() {
		return "MWCNR";
	}
	
	private int nextInt() {
		seed = multiplier * ((int)seed) + (seed >>> 32);
		return (int)seed;
	}
	
	public long nextLong() {
		long left  = nextInt();
		long right = nextInt();
		return (left << 32) ^ right;
	}
	
    public long next64bits() {
    	return nextLong();
    }
	
	public double nextDouble() {
		return Math.abs( (double)nextInt() / Integer.MAX_VALUE );
	}
	
}

File: python-implementation/src/pnrg/LCGs/MWC.py
import numpy as np

from pnrg import RNG

def lrshift(val, n) -> np.int64:
    """
    Logically right shifts value by n places.
    :param val: value to shift
    :param n: number of bits to shift by.
    :return: result of right shift.
    """
    return (val % (1 << 64)) >> n

class MWC(RNG):
    """
    An implementation of the Multiply-With-Carry generator from Numerical Recipes
    """
    def __init__(self, seed):
        super().__init__("MWCNR")
        self.seed = np.int64(seed)
        self.multiplier = np.int64(0xffffda61)

    def next_32_int(self) -> np.int32:
        self.seed = self.multiplier * (np.int32(self.seed) + lrshift(self.seed, 32))
        return np.int32(self.seed)

    def next_int(self) -> np.int64:
        left = np.int64(self.next_32_int())
        right = np.int64(self.next_32_int())
        return (left << 32) ^ right

    def next_float(self) -> np.float64:
        return abs(self.next_int() / np.iinfo(np.int64).max)

    def next_64_bits(self) -> np.int64:
        return self.next_int()